### PR TITLE
fix: reject data urls in uploadFileFromUrl()

### PR DESCRIPTION
**Pull request by markflorkowski**
- **reject data urls**
- **add test**

### CODE REVIEW

**Review by juliusmarminge — APPROVED** [review]

File: packages/uploadthing/src/sdk/index.ts [hunk 1]
```diff
@@ -219,7 +219,19 @@ export class UTApi {
...
+
+        if (typeof url === "string") {
+          // since dataurls will result in name being too long, tell the user
+          // to use uploadFiles instead.
+          if (url.startsWith("data:")) {
+            throw new UploadThingError({
+              code: "BAD_REQUEST",
+              message:
+                "Please use uploadFiles() for data URLs. uploadFilesFromUrl() is intended for use with remote URLs only.",
+            });
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:231]: wait should we throw? the inner function returns either data or error and let's the other continue

Comment (markflorkowski) [reply packages/uploadthing/src/sdk/index.ts:231]: We throw below if ant of the files fail to download. Possibly we should not throw in either case then...

Comment (juliusmarminge) [reply packages/uploadthing/src/sdk/index.ts:231]: right... we should probably just make sure to add an error for that place in the final result array?

File: packages/uploadthing/src/sdk/index.ts [hunk 2]
```diff
@@ -256,7 +287,9 @@ export class UTApi {
       },
     );
 
-    const uploadFileResponse = Array.isArray(urls) ? uploads : uploads[0];
+    const uploadFileResponse = Array.isArray(urls)
+      ? [...errors, ...uploads]
+      : [...errors, ...uploads][0];
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:292]: Would be nice if order was preserved

File: packages/uploadthing/src/sdk/sdk.test.ts [hunk 3]
```diff
@@ -135,6 +135,36 @@ describe("uploadFilesFromUrl", () => {
...
+        return Promise.resolve(Response.json([]));
+      },
+    });
+
+    const result = await utapi.uploadFilesFromUrl(
+      "data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==",
+    );
+    expect(result).toMatchInlineSnapshot(
+      `[Error: Please use uploadFiles() for data URLs. uploadFilesFromUrl() is intended for use with remote URLs only.]`,
+    );
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/sdk.test.ts:166]: `{ data: null, error }`?

File: packages/uploadthing/src/sdk/index.ts (hunk 2 above)
Comment (markflorkowski) [reply packages/uploadthing/src/sdk/index.ts:292]: Would require infra changes, and I'm not sure how the UploadThingError being passed over the wire would go

Comment (juliusmarminge) [reply packages/uploadthing/src/sdk/index.ts:292]: Would it? We'd just need to keep track of what indexes failed?

File: packages/uploadthing/src/sdk/index.ts [hunk 4]
```diff
@@ -228,24 +243,25 @@ export class UTApi {
...
+          return undefined;
         }
         logger.debug("Finished downloading file. Reading blob...");
         const blob = await fileResponse.blob();
         logger.debug("Finished reading blob.");
-        return Object.assign(blob, { name: filename });
+        return new UTFile([blob], filename);
       }),
-    );
+    ).then((files) => files.filter((x): x is UTFile => x !== undefined));
```
Comment (markflorkowski) [inline packages/uploadthing/src/sdk/index.ts:258]:
```suggestion
    ).then((files) => files.filter(Boolean);
```

?

File: packages/uploadthing/test/sdk.test.ts [hunk 5]
```diff
@@ -212,6 +212,61 @@ describe("uploadFilesFromUrl", () => {
       },
     );
   });
+
+  // if passed data url, array contains UploadThingError
+  it("return is error if data url is passed", async () => {
```
Comment (markflorkowski) [inline packages/uploadthing/test/sdk.test.ts:217]:
```suggestion
  it("returns error if data url is passed", async () => {
```

**Pull request merged**